File: frontend/lib/models/reminder.dart
class Reminder {
  final String id;
  final String? categoryId;
  final String title;
  final String? description;
  final String priority;
  final String triggerType;
  final Map<String, dynamic> triggerConfig;
  final int advanceNotice;
  final String? repeatRule;
  final String status;
  final DateTime createdAt;

  Reminder({
    required this.id,
    this.categoryId,
    required this.title,
    this.description,
    required this.priority,
    required this.triggerType,
    required this.triggerConfig,
    this.advanceNotice = 0,
    this.repeatRule,
    required this.status,
    required this.createdAt,
  });

  factory Reminder.fromJson(Map<String, dynamic> json) => Reminder(
    id: json['id'],
    categoryId: json['category_id'],
    title: json['title'],
    description: json['description'],
    priority: json['priority'],
    triggerType: json['trigger_type'],
    triggerConfig: Map<String, dynamic>.from(json['trigger_config'] ?? {}),
    advanceNotice: json['advance_notice'] ?? 0,
    repeatRule: json['repeat_rule'],
    status: json['status'],
    createdAt: DateTime.parse(json['created_at']),
  );

  Map<String, dynamic> toJson() => {
    'category_id': categoryId,
    'title': title,
    'description': description,
    'priority': priority,
    'trigger_type': triggerType,
    'trigger_config': triggerConfig,
    'advance_notice': advanceNotice,
    'repeat_rule': repeatRule,
  };
}

File: backend/app/models/reminder.py
import uuid
from sqlalchemy import Column, String, ForeignKey, Text, Integer, DateTime
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.database import Base


class Reminder(Base):
    __tablename__ = "reminders"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(String(36), ForeignKey("users.id"), nullable=False)
    category_id = Column(String(36), ForeignKey("categories.id"), nullable=True)
    title = Column(String(200), nullable=False)
    description = Column(Text, nullable=True)
    priority = Column(String(20), default="medium")
    trigger_type = Column(String(50), nullable=False)
    trigger_config = Column(Text, nullable=False, default="{}")
    advance_notice = Column(Integer, default=0)
    repeat_rule = Column(String(50), nullable=True)
    status = Column(String(20), default="active")
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())

    user = relationship("User", back_populates="reminders")
    category = relationship("Category", back_populates="reminders")
    event_trigger = relationship("EventTrigger", back_populates="reminder", uselist=False, cascade="all, delete-orphan")
    histories = relationship("ReminderHistory", back_populates="reminder", cascade="all, delete-orphan")


class EventTrigger(Base):
    __tablename__ = "event_triggers"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    reminder_id = Column(String(36), ForeignKey("reminders.id"), nullable=False, unique=True)
    event_type = Column(String(50), nullable=False)
    config = Column(Text, nullable=False, default="{}")
    last_checked = Column(DateTime(timezone=True), nullable=True)
    last_triggered = Column(DateTime(timezone=True), nullable=True)

    reminder = relationship("Reminder", back_populates="event_trigger")
